### ISSUE TITLE
Rule Proposal: guard-clause

### ISSUE THREAD

**Issue by Kerrick** [eslint/eslint#10858]
**Please describe what the rule should do:** The rule should enforce the Guard Clause Pattern.

Documentation on this pattern:

- https://en.wikipedia.org/wiki/Guard_(computer_science)
- https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html
- http://wiki.c2.com/?GuardClause

Prior art:

- [Rubocop: Style/GuardClause](http://docs.rubocop.org/en/latest/cops_style/#styleguardclause)

**What category of rule is this? (place an "X" next to just one item)**

[X] Enforces code style
[ ] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
function computeThing(thing) {
  if (thing.foo) {
    updateWithBar(thing.foo);
    recompute(thing);
    // ...
  }
}
// Should Become
function computeThing(thing) {
  if (!thing.foo) { return; }
  updateWithBar(thing.foo);
  recompute(thing);
  // ...
}
```

```js
for (let foo of bar) {
  if (foo.price > 10) {
    foo.putOnSale();
    foo.advertise();
    // ...
  }
}
// Should Become
for (let foo of bar) {
  if (foo.price <= 10) { continue; }
  foo.putOnSale();
  foo.advertise();
  // ...
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**

It is an old enough, common enough pattern that other RuboCopy has included it in their core rules. It is also of a similar ethos to `no-else-return`.
